### PR TITLE
drivers: timer: apic_tsc: add support for read HW clock cycle per sec on runtime 

### LINKED ISSUE #69318: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is not right for acrn_defconfig
**Describe the bug**
defconfig for x86 acrn boards declares:
```
# boards/x86/acrn/acrn_defconfig
CONFIG_APIC_TSC_DEADLINE_TIMER=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=25000000
```
This is definitely not right as TSC frequency is the CPU Core Freq which should not be 25MHz small.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? x86
 - What have you tried to diagnose or workaround this issue? Runing a sleep test app.

**Impact**

Time/timeout not work correctly.

### PR DESCRIPTION
**Pull request by najumon1980**
Added support for read TSC timer clock frequency (HW clock cycle per sec) on runtime using cpuid. TSC APIC timer frequency will be the processor’s bus clock or core crystal clock frequency divided by the value specified in the divide configuration register. But core clock frequency varies from platform to platform and even with different SKUs (for Intel platforms it usually from ~1GHz to ~4GHz). But even with same platform type (for different SKU) this can be changed. So best way is read via CPUID (leaf 0x15) and then configure dynamically instead of using statically (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)

### CODE REVIEW

**Review by cfriedt — DISMISSED** [review]
Lgtm - maybe a rebase?

**Comment by jhedberg** [comment]
@najumon1980 this needs rebasing

**Comment by najumon1980** [comment]
The CI job one failed looks failing in mainline as well and hence it not introduce by this PR

**Review by jhedberg — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

File: drivers/timer/apic_tsc.c [hunk 1]
```diff
@@ -174,6 +182,23 @@ static int sys_clock_driver_init(void)
...
+
+	eax = 0;
+	ebx = 0;
+	ecx = 0;
+	if (!__get_cpuid(0x15, &eax, &ebx, &ecx, &edx)) {
+		__ASSERT(false, "error cpuid leaf 0x15\n");
+	}
+
+	tsc_freq = ecx * ebx/eax;
+	__ASSERT((ebx != 0) && (ecx != 0), "No TSC/core crystal clock ratio is enumerated");
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply drivers/timer/apic_tsc.c:195]: updated

**Review by jhedberg — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Comment by najumon1980** [comment]
@sudhirbrnwl could you please approve if no more comment ?

**Review by sudhirbrnwl — APPROVED** [review]

**Comment by sudhirbrnwl** [comment]
> @sudhirbrnwl could you please approve if no more comment ?

Done

**Comment by najumon1980** [comment]
@nashif is this PR ok to merge if no more approval required ?

**Comment by najumon1980** [comment]
@andyross could you please approve if no more comments ?

File: drivers/timer/apic_tsc.c [hunk 2]
```diff
@@ -174,6 +182,23 @@ static int sys_clock_driver_init(void)
...
+	eax = 0;
+	ebx = 0;
+	ecx = 0;
+	if (!__get_cpuid(0x15, &eax, &ebx, &ecx, &edx)) {
+		__ASSERT(false, "error cpuid leaf 0x15\n");
+	}
+
+	__ASSERT((ebx != 0) && (ecx != 0), "No TSC/core crystal clock ratio is enumerated");
+
+	tsc_freq = ecx * ebx/eax;
```
Comment (andyross) [inline drivers/timer/apic_tsc.c:197]: These are 32 bit values, and the precedence here is going to compute (ecx * ebx) / eax.  Does the architecture guarantee that the numerator can't overflow?  Probably best to do the multiply in 64 bit.

File: include/zephyr/sys_clock.h [hunk 3]
```diff
@@ -146,11 +146,14 @@ typedef struct {
 
 /** @endcond */
 
-#if defined(CONFIG_SYS_CLOCK_EXISTS) && \
-	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC == 0)
+#if defined(CONFIG_SYS_CLOCK_EXISTS) && !defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME)
+
+#if (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC == 0)
 #error "SYS_CLOCK_HW_CYCLES_PER_SEC must be non-zero!"
```
Comment (andyross) [inline include/zephyr/sys_clock.h:152]: If you're mucking with the logic here anyway, BUILD_ASSERT() is the more modern way to do this sort of thing.

File: drivers/timer/apic_tsc.c (hunk 2 above)
Comment (najumon1980) [reply drivers/timer/apic_tsc.c:197]: Current Zephyr supported Intel platform limited to maximum 2.4GHz. The z_clock_hw_cycles_per_sec is also 32bit variable currently

**Review by andyross — COMMENTED** [review]
One more tiny nitpick, nonfatal.  Basically I'm just worried about precision. If the answer is "the architecture guarantees that the product is less than UINT_MAX" then that's fine and assume a +1 from me.  Otherwise just add a cast to uint64_t and likewise assume +1.

File: drivers/timer/apic_tsc.c (hunk 2 above)
Comment (andyross) [reply drivers/timer/apic_tsc.c:197]: The **result** is representable in 32 bits.  But the intermediate product computed before the division may not be.  Add a cast and do it in 64 bits.

**Review by andyross — CHANGES_REQUESTED** [review]
I looked it up.  I'm too lazy to dig through the IAM (why is that not indexed online?  PDFs are great and all but it's like 12k pages now!) but felixcloutier says:

> If EBX[31:0] is 0, the TSC/”core crystal clock” ratio is not enumerated. EBX[31:0]/EAX[31:0] indicates the ratio of the TSC frequency and the core crystal clock frequency. If ECX is 0, the nominal core crystal clock frequency is not enumerated. “TSC frequency” = “core crystal clock frequency” * EBX/EAX. The core crystal clock may differ from the reference clock, bus clock, or core clock frequencies. EAX Bits 31-00: An unsigned integer which is the denominator of the TSC/”core crystal clock” ratio. EBX Bits 31-00: An unsigned integer which is the numerator of the TSC/”core crystal clock” ratio. ECX Bits 31-00: An unsigned integer which is the nominal frequency of the core crystal clock in Hz. EDX Bits 31-00: Reserved = 0.

So indeed, I don't see any phrasing about overflow protection.  You need to provide it yourself.  Just something like this should be fine:

```
tsc_freq = (uint32_t)((ecx * (uint64_t)ebx) / eax);
```

(Extra parens added because the MISRA police are patrolling the block
again and this would otherwise be flagged AIUI).

As before: assume +1 from me if that change is made and you can't find me to approve directly.

**Comment by npitre** [comment]
> Extra parens added because the MISRA police are patrolling the block again

LOL!

Made my day.

File: drivers/timer/apic_tsc.c (hunk 2 above)
Comment (najumon1980) [reply drivers/timer/apic_tsc.c]: updated

File: include/zephyr/sys_clock.h (hunk 3 above)
Comment (najumon1980) [reply include/zephyr/sys_clock.h]: updated

**Review by andyross — APPROVED** [review]
Thanks

**Review by jhedberg — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
@najumon1980 looks like CI broke after merging this,  7c03e5de7f0298e9219f82da52b75696c45baa9d specifically, can you look into it? the test is `
west build -t run -p -b qemu_x86 -T tests/subsys/logging/log_timestamp/logging.output.custom_timestamp`

**Comment by najumon1980** [comment]
> @najumon1980 looks like CI broke after merging this, [7c03e5d](https://github.com/zephyrproject-rtos/zephyr/commit/7c03e5de7f0298e9219f82da52b75696c45baa9d) specifically, can you look into it? the test is ` west build -t run -p -b qemu_x86 -T tests/subsys/logging/log_timestamp/logging.output.custom_timestamp`

hi Johan, could you please help me to point to CI job ?

**Comment by jhedberg** [comment]
> > @najumon1980 looks like CI broke after merging this, [7c03e5d](https://github.com/zephyrproject-rtos/zephyr/commit/7c03e5de7f0298e9219f82da52b75696c45baa9d) specifically, can you look into it? the test is ` west build -t run -p -b qemu_x86 -T tests/subsys/logging/log_timestamp/logging.output.custom_timestamp`
> 
> hi Johan, could you please help me to point to CI job ?

@najumon1980 I'm not sure what you mean, but I think it should be possible to debug and fix this simply using the command line that @fabiobaltieri provided. I was able to verify myself that the test fails with current main branch but passes with https://github.com/zephyrproject-rtos/zephyr/commit/7c03e5de7f0298e9219f82da52b75696c45baa9d reverted.

**Comment by jhedberg** [comment]
@najumon1980 btw, I did some quick debugging myself, but didn't get very far. One thing I did verify is that at least `sys_clock_hw_cycles_per_sec()` returns the same value both before and after the problematic commit.

**Comment by najumon1980** [comment]
It seems default TIMER_READS_ITS_FREQUENCY_AT_RUNTIME enabled for HPET timer when x86 platform selected and for QEMU/KVM the CPUID is not supported.  Just created following draft PR for the fix:  https://github.com/zephyrproject-rtos/zephyr/pull/72336

**Comment by aescolar** [comment]
> @najumon1980 looks like CI broke after merging this, [7c03e5d](https://github.com/zephyrproject-rtos/zephyr/commit/7c03e5de7f0298e9219f82da52b75696c45baa9d) specifically, can you look into it? the test is ` west build -t run -p -b qemu_x86 -T tests/subsys/logging/log_timestamp/logging.output.custom_timestamp`

Thanks @fabiobaltieri issue created: https://github.com/zephyrproject-rtos/zephyr/issues/72344

**Comment by najumon1980** [comment]
Looks sys_clock_hw_cycles_per_sec() used in log_core_init() which invoked during cstart (EARLY init)  but HPET driver init is PRE_KERNEL_2.  So the previous code where z_clock_hw_cycles_per_sec need to init with default CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC (though it is not right).

**Pull request merged**